go/test/endtoend/cluster_test.go

```diff
@@ -439,6 +441,44 @@ func checkHeartbeatsWritten(multipoolerAddr string) (bool, error) {
 	return count > 0, nil
 }
 
+// findReadyMultigateway finds a multigateway that is ready to execute queries.
+// It tries each provided port and returns the first one that can successfully execute
+// a query. This is necessary because:
+//  1. PoolerDiscovery is async and may not have discovered poolers yet
+//  2. lib/pq Ping uses an empty query (";") that bypasses pooler discovery
+//  3. Each multigateway only discovers poolers in its own zone, and the PRIMARY
+//     may be in any zone after bootstrap
+func findReadyMultigateway(t *testing.T, pgPorts []int, timeout time.Duration) (int, error) {
```

Comment:
Feels like this should be taking a context instead of a duration for timeout to make sure it respects t.Context() getting cancelled if the test was started with a timeout, etc.